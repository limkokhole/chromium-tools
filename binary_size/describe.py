# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
"""Methods for converting model objects to human-readable formats."""

import datetime
import itertools
import time

import models


def _PrettySize(size):
  # Arbitrarily chosen cut-off.
  if abs(size) < 2000:
    return '%d bytes' % size
  # Always show 3 digits.
  size /= 1024.0
  if abs(size) < 10:
    return '%.2fkb' % size
  elif abs(size) < 100:
    return '%.1fkb' % size
  elif abs(size) < 1024:
    return '%dkb' % size
  size /= 1024.0
  if abs(size) < 10:
    return '%.2fmb' % size
  # We shouldn't be seeing sizes > 100mb.
  return '%.1fmb' % size


class Describer(object):
  def __init__(self, verbose=False):
    self.verbose = verbose

  def _DescribeSectionSizes(self, section_sizes):
    relevant_names = models.SECTION_TO_SECTION_NAME.values()
    section_names = sorted(k for k in section_sizes.iterkeys()
                           if k in relevant_names or k.startswith('.data'))
    total_bytes = sum(v for k, v in section_sizes.iteritems()
                      if k in section_names and k != '.bss')
    yield ''
    yield 'Section Sizes (Total={:,} bytes):'.format(total_bytes)
    for name in section_names:
      size = section_sizes[name]
      if name == '.bss':
        yield '    {}: {:,} bytes (not included in totals)'.format(name, size)
      else:
        percent = float(size) / total_bytes if total_bytes else 0
        yield '    {}: {:,} bytes ({:.1%})'.format(name, size, percent)

    if self.verbose:
      yield ''
      yield 'Other section sizes:'
      section_names = sorted(k for k in section_sizes.iterkeys()
                             if k not in section_names)
      for name in section_names:
        yield '    {}: {:,} bytes'.format(name, section_sizes[name])

  def _DescribeSymbol(self, sym):
    # SymbolGroups are passed here when we don't want to expand them.
    if sym.IsGroup():
      if self.verbose:
        yield ('{} {:<8} {} (count={})  padding={}  '
               'size_without_padding={}').format(
                    sym.section, sym.size, sym.name, len(sym), sym.padding,
                    sym.size_without_padding)
      else:
        yield '{} {:<8} {} (count={})'.format(sym.section, sym.size, sym.name,
                                              len(sym))
      return

    if self.verbose:
      yield '{}@0x{:<8x}  size={}  padding={}  size_without_padding={}'.format(
          sym.section, sym.address, sym.size, sym.padding,
          sym.size_without_padding)
      yield '    source_path={} \tobject_path={}'.format(
          sym.source_path, sym.object_path)
      if sym.full_name:
        yield '    is_anonymous={}  full_name={}'.format(
            int(sym.is_anonymous), sym.full_name)
      if sym.name:
        yield '    is_anonymous={}  name={}'.format(
            int(sym.is_anonymous), sym.name)
    else:
      yield '{}@0x{:<8x}  {:<7} {}'.format(
          sym.section, sym.address, sym.size,
          sym.source_path or sym.object_path or '{no path}')
      if sym.name:
        yield '    {}'.format(sym.name)

  def _DescribeSymbolGroup(self, group, prefix_func=None):
    total_size = group.size
    yield 'Showing {:,} symbols with total size: {} bytes'.format(
        len(group), total_size)
    code_syms = group.WhereInSection('t')
    code_size = code_syms.size
    ro_size = code_syms.Inverted().WhereInSection('r').size
    yield '.text={:<10} .rodata={:<10} other={:<10} total={}'.format(
        _PrettySize(code_size), _PrettySize(ro_size),
        _PrettySize(total_size - code_size - ro_size),
        _PrettySize(total_size))
    unique_paths = set(s.object_path for s in group)
    yield 'Number of object files: {}'.format(len(unique_paths))
    yield ''
    yield 'First columns are: running total, type, size'

    running_total = 0
    prefix = ''
    sorted_syms = group if group.is_sorted else group.Sorted()

    prefix = ''
    for s in sorted_syms:
      if group.IsBss() or not s.IsBss():
        running_total += s.size
      for l in self._DescribeSymbol(s):
        if l[:4].isspace():
          yield '{} {}'.format(' ' * (8 + len(prefix)), l)
        else:
          if prefix_func:
            prefix = prefix_func(s)
          yield '{}{:8} {}'.format(prefix, running_total, l)

  def _DescribeSymbolDiff(self, diff):
    header_template = ('{} symbols added (+), {} changed (~), {} removed (-), '
                       '{} unchanged ({})')
    unchanged_msg = '=' if self.verbose else 'not shown'
    symbol_delta_desc = [header_template.format(
        diff.added_count, diff.changed_count, diff.removed_count,
        diff.unchanged_count, unchanged_msg)]

    similar_paths = set()
    added_paths = set()
    removed_paths = set()
    for s in diff:
      if diff.IsAdded(s):
        added_paths.add(s.object_path)
      elif diff.IsRemoved(s):
        removed_paths.add(s.object_path)
      else:
        similar_paths.add(s.object_path)
    added_paths.difference_update(similar_paths)
    removed_paths.difference_update(similar_paths)
    path_delta_desc = ['{} object files added, {} removed'.format(
        len(added_paths), len(removed_paths))]
    if self.verbose and len(added_paths):
      path_delta_desc.append('Added files:')
      path_delta_desc.extend('  ' + p for p in sorted(added_paths))
    if self.verbose and len(removed_paths):
      path_delta_desc.append('Removed files:')
      path_delta_desc.extend('  ' + p for p in sorted(removed_paths))

    def prefix_func(sym):
      if diff.IsAdded(sym):
        return '+ '
      if diff.IsRemoved(sym):
        return '- '
      if sym.size:
        return '~ '
      return '= '

    diff = diff if self.verbose else diff.WhereNotUnchanged()
    group_desc = self._DescribeSymbolGroup(diff, prefix_func=prefix_func)
    return itertools.chain(symbol_delta_desc, path_delta_desc, ('',),
                           group_desc)

  def _DescribeSizeInfoDiff(self, diff):
    common_metadata = {k: v for k, v in diff.old_metadata.iteritems()
                       if diff.new_metadata[k] == v}
    old_metadata = {k: v for k, v in diff.old_metadata.iteritems()
                    if k not in common_metadata}
    new_metadata = {k: v for k, v in diff.new_metadata.iteritems()
                    if k not in common_metadata}
    metadata_desc = itertools.chain(
        ('Common Metadata:',),
        ('    %s' % line for line in DescribeMetadata(common_metadata)),
        ('Old Metadata:',),
        ('    %s' % line for line in DescribeMetadata(old_metadata)),
        ('New Metadata:',),
        ('    %s' % line for line in DescribeMetadata(new_metadata)))
    section_desc = self._DescribeSectionSizes(diff.section_sizes)
    group_desc = self.GenerateLines(diff.symbols)
    return itertools.chain(metadata_desc, section_desc, ('',), group_desc)

  def _DescribeSizeInfo(self, size_info):
    metadata_desc = itertools.chain(
        ('Metadata:',),
        ('    %s' % line for line in DescribeMetadata(size_info.metadata)))
    section_desc = self._DescribeSectionSizes(size_info.section_sizes)
    coverage_desc = ()
    if self.verbose:
      coverage_desc = itertools.chain(
          ('',), DescribeSizeInfoCoverage(size_info))
    group_desc = self.GenerateLines(size_info.symbols)
    return itertools.chain(metadata_desc, section_desc, coverage_desc, ('',),
                           group_desc)

  def GenerateLines(self, obj):
    if isinstance(obj, models.SizeInfoDiff):
      return self._DescribeSizeInfoDiff(obj)
    if isinstance(obj, models.SizeInfo):
      return self._DescribeSizeInfo(obj)
    if isinstance(obj, models.SymbolDiff):
      return self._DescribeSymbolDiff(obj)
    if isinstance(obj, models.SymbolGroup):
      return self._DescribeSymbolGroup(obj)
    if isinstance(obj, models.Symbol):
      return self._DescribeSymbol(obj)
    return (repr(obj),)


def DescribeSizeInfoCoverage(size_info):
  """Yields lines describing how accurate |size_info| is."""
  for section in models.SECTION_TO_SECTION_NAME:
    if section == 'd':
      expected_size = sum(v for k, v in size_info.section_sizes.iteritems()
                          if k.startswith('.data'))
    else:
      expected_size = size_info.section_sizes[
          models.SECTION_TO_SECTION_NAME[section]]

    def one_stat(group):
      template = ('Section %s has %.1f%% of %d bytes accounted for from '
                  '%d symbols. %d bytes are unaccounted for. Padding '
                  'accounts for %d bytes')
      actual_size = group.size
      count = len(group)
      padding = group.padding
      size_percent = 100.0 * actual_size / expected_size
      return (template % (section, size_percent, actual_size, count,
                          expected_size - actual_size, padding))

    in_section = size_info.symbols.WhereInSection(section)
    yield one_stat(in_section)

    star_syms = in_section.WhereNameMatches(r'^\*')
    attributed_syms = star_syms.Inverted().WhereHasAnyAttribution()
    anonymous_syms = attributed_syms.Inverted()
    if star_syms or anonymous_syms:
      missing_size = star_syms.size + anonymous_syms.size
      yield ('+ Without %d merge sections and %d anonymous entries ('
                  'accounting for %d bytes):') % (
          len(star_syms),  len(anonymous_syms), missing_size)
      yield '+ ' + one_stat(attributed_syms)


def _UtcToLocal(utc):
  epoch = time.mktime(utc.timetuple())
  offset = (datetime.datetime.fromtimestamp(epoch) -
            datetime.datetime.utcfromtimestamp(epoch))
  return utc + offset


def DescribeMetadata(metadata):
  display_dict = metadata.copy()
  timestamp = display_dict.get(models.METADATA_ELF_MTIME)
  if timestamp:
    timestamp_obj = datetime.datetime.utcfromtimestamp(timestamp)
    display_dict[models.METADATA_ELF_MTIME] = (
        _UtcToLocal(timestamp_obj).strftime('%Y-%m-%d %H:%M:%S'))
  gn_args = display_dict.get(models.METADATA_GN_ARGS)
  if gn_args:
    display_dict[models.METADATA_GN_ARGS] = '; '.join(gn_args)
  return sorted('%s=%s' % t for t in display_dict.iteritems())


def GenerateLines(obj, verbose=False):
  return Describer(verbose).GenerateLines(obj)


def WriteLines(lines, func):
  for l in lines:
    func(l)
    func('\n')
