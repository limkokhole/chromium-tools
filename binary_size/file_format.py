# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""Deals with loading & saving .size files."""

import ast
import gzip
import models


# File format version for .size files.
_SERIALIZATION_VERSION = 1


def EndsWithMaybeGz(path, suffix):
  return path.endswith(suffix) or path.endswith(suffix + '.gz')


def OpenMaybeGz(path, mode=None):
  """Calls `gzip.open()` if |path| ends in ".gz", otherwise calls `open()`."""
  if path.endswith('.gz'):
    if mode and 'w' in mode:
      return gzip.GzipFile(path, mode, 1)
    return gzip.open(path, mode)
  return open(path, mode or 'r')


def _SaveSizeInfoToFile(result, file_obj):
  """Saves the result to the given file object."""
  # Store one bucket per line.
  file_obj.write('%d\n' % _SERIALIZATION_VERSION)
  file_obj.write('%r\n' % result.section_sizes)
  file_obj.write('%d\n' % len(result.symbols))
  prev_section_name = None
  # Store symbol fields as tab-separated.
  # Store only non-derived fields.
  for symbol in result.symbols:
    if symbol.section_name != prev_section_name:
      file_obj.write('%s\n' % symbol.section_name)
      prev_section_name = symbol.section_name
    # Don't write padding nor name since these are derived values.
    file_obj.write('%x\t%x\t%s\t%s\n' % (
        symbol.address, symbol.size_without_padding,
        symbol.function_signature or symbol.name, symbol.path))


def _LoadSizeInfoFromFile(file_obj):
  """Loads a result from the given file."""
  lines = iter(file_obj)
  actual_version = int(next(lines))
  assert actual_version == _SERIALIZATION_VERSION, (
      'Version mismatch. Need to write some upgrade code.')

  section_sizes = ast.literal_eval(next(lines))
  num_syms = int(next(lines))
  symbol_list = [None] * num_syms
  section_name = None
  for i in xrange(num_syms):
    line = next(lines)[:-1]
    if '\t' not in line:
      section_name = line
      line = next(lines)[:-1]
    new_sym = models.Symbol.__new__(models.Symbol)
    parts = line.split('\t')
    new_sym.section_name = section_name
    new_sym.address = int(parts[0], 16)
    new_sym.size = int(parts[1], 16)
    new_sym.name = parts[2]
    new_sym.path = parts[3]
    new_sym.padding = 0  # Derived
    new_sym.function_signature = None  # Derived
    symbol_list[i] = new_sym

  return models.SizeInfo(models.SymbolGroup(symbol_list), section_sizes)


def SaveSizeInfo(result, path):
  with OpenMaybeGz(path, 'wb') as f:
    _SaveSizeInfoToFile(result, f)


def LoadSizeInfo(path):
  with OpenMaybeGz(path) as f:
    return _LoadSizeInfoFromFile(f)
