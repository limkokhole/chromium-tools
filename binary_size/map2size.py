#!/usr/bin/env python
# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""Main Python API for analyzing binary size."""

import argparse
import datetime
import distutils.spawn
import gzip
import logging
import os
import re
import subprocess
import sys

import describe
import file_format
import function_signature
import helpers
import linker_map_parser
import models
import ninja_parser


def _OpenMaybeGz(path, mode=None):
  """Calls `gzip.open()` if |path| ends in ".gz", otherwise calls `open()`."""
  if path.endswith('.gz'):
    if mode and 'w' in mode:
      return gzip.GzipFile(path, mode, 1)
    return gzip.open(path, mode)
  return open(path, mode or 'r')


def _UnmangleRemainingSymbols(symbol_group, tool_prefix):
  """Uses c++filt to unmangle any symbols that need it."""
  to_process = [s for s in symbol_group if s.name.startswith('_Z')]
  if not to_process:
    return

  logging.info('Unmangling %d names', len(to_process))
  proc = subprocess.Popen([tool_prefix + 'c++filt'], stdin=subprocess.PIPE,
                          stdout=subprocess.PIPE)
  stdout = proc.communicate('\n'.join(s.name for s in to_process))[0]
  assert proc.returncode == 0

  for i, line in enumerate(stdout.splitlines()):
    to_process[i].name = line


def _NormalizeNames(symbol_group):
  """Ensures that all names are formatted in a useful way.

  This includes:
    - Assigning of |full_name|.
    - Stripping of return types in |full_name| and |name| (for functions).
    - Stripping parameters from |name|.
    - Moving "vtable for" and the like to be suffixes rather than prefixes.
  """
  found_prefixes = set()
  for symbol in symbol_group:
    if symbol.name.startswith('*'):
      # See comment in _RemoveDuplicatesAndCalculatePadding() about when this
      # can happen.
      continue

    # E.g.: vtable for FOO
    idx = symbol.name.find(' for ', 0, 30)
    if idx != -1:
      found_prefixes.add(symbol.name[:idx + 4])
      symbol.name = symbol.name[idx + 5:] + ' [' + symbol.name[:idx] + ']'

    # E.g.: virtual thunk to FOO
    idx = symbol.name.find(' to ', 0, 30)
    if idx != -1:
      found_prefixes.add(symbol.name[:idx + 3])
      symbol.name = symbol.name[idx + 4:] + ' [' + symbol.name[:idx] + ']'

    # Strip out return type, and identify where parameter list starts.
    if symbol.section == 't':
      symbol.full_name, symbol.name = function_signature.Parse(symbol.name)

    # Remove anonymous namespaces (they just harm clustering).
    non_anonymous = symbol.name.replace('(anonymous namespace)::', '')
    if symbol.name != non_anonymous:
      symbol.is_anonymous = True
      symbol.name = non_anonymous
      symbol.full_name = symbol.full_name.replace(
          '(anonymous namespace)::', '')

    if symbol.section != 't' and '(' in symbol.name:
      # Pretty rare. Example:
      # blink::CSSValueKeywordsHash::findValueImpl(char const*)::value_word_list
      symbol.full_name = symbol.name
      symbol.name = re.sub(r'\(.*\)', '', symbol.full_name)

  logging.debug('Found name prefixes of: %r', found_prefixes)


def _NormalizeObjectPaths(symbol_group):
  """Ensures that all paths are formatted in a useful way."""
  for symbol in symbol_group:
    path = symbol.object_path
    if path.startswith('obj/'):
      # Convert obj/third_party/... -> third_party/...
      path = path[4:]
    elif path.startswith('../../'):
      # Convert ../../third_party/... -> third_party/...
      path = path[6:]
    if path.endswith(')'):
      # Convert foo/bar.a(baz.o) -> foo/bar.a/(baz.o)
      start_idx = path.index('(')
      path = os.path.join(path[:start_idx], path[start_idx:])
    symbol.object_path = path


def _NormalizeSourcePath(path):
  if path.startswith('gen/'):
    # Convert gen/third_party/... -> third_party/...
    return path[4:]
  if path.startswith('../../'):
    # Convert ../../third_party/... -> third_party/...
    return path[6:]
  return path


def _ExtractSourcePaths(symbol_group, output_directory):
  """Fills in the .source_path attribute of all symbols."""
  mapper = ninja_parser.SourceFileMapper(output_directory)

  for symbol in symbol_group:
    object_path = symbol.object_path
    if symbol.source_path or not object_path:
      continue
    # We don't have source info for prebuilt .a files.
    if not object_path.startswith('..'):
      source_path = mapper.FindSourceForPath(object_path)
      if source_path:
        symbol.source_path = _NormalizeSourcePath(source_path)
      else:
        logging.warning('Could not find source path for %s', object_path)
  logging.debug('Parsed %d .ninja files.', mapper.GetParsedFileCount())


def _RemoveDuplicatesAndCalculatePadding(symbol_group):
  """Removes symbols at the same address and calculates the |padding| field.

  Symbols must already be sorted by |address|.
  """
  to_remove = set()
  all_symbols = symbol_group.symbols
  for i, symbol in enumerate(all_symbols[1:]):
    prev_symbol = all_symbols[i]
    if prev_symbol.section_name != symbol.section_name:
      continue
    if symbol.address > 0 and prev_symbol.address > 0:
      # Fold symbols that are at the same address (happens in nm output).
      if symbol.address == prev_symbol.address:
        symbol.size = max(prev_symbol.size, symbol.size)
        to_remove.add(i + 1)
        continue
      # Even with symbols at the same address removed, overlaps can still
      # happen. In this case, padding will be negative (and this is fine).
      padding = symbol.address - prev_symbol.end_address
      # These thresholds were found by manually auditing arm32 Chrome.
      # E.g.: Set them to 0 and see what warnings get logged.
      # TODO(agrieve): See if these thresholds make sense for architectures
      #     other than arm32.
      if (symbol.section in 'rd' and padding >= 256 or
          symbol.section in 't' and padding >= 64):
        # For nm data, this is caused by data that has no associated symbol.
        # The linker map file lists them with no name, but with a file.
        # Example:
        #   .data 0x02d42764 0x120 .../V8SharedWorkerGlobalScope.o
        # Where as most look like:
        #   .data.MANGLED_NAME...
        logging.debug('Large padding of %d between:\n  A) %r\n  B) %r' % (
                      padding, prev_symbol, symbol))
        continue
      symbol.padding = padding
      symbol.size += padding
      assert symbol.size >= 0, 'Symbol has negative size: ' + (
          '%r\nprev symbol: %r' % (symbol, prev_symbol))
  # Map files have no overlaps, so worth special-casing the no-op case.
  if to_remove:
    logging.info('Removing %d overlapping symbols', len(to_remove))
    symbol_group.symbols = (
        [s for i, s in enumerate(all_symbols) if i not in to_remove])


def AddOptions(parser):
  parser.add_argument('--tool-prefix', default='',
                      help='Path prefix for c++filt.')
  parser.add_argument('--output-directory',
                      help='Path to the root build directory.')


def _DetectToolPrefix(tool_prefix, input_file, output_directory=None):
  """Detects values for --tool-prefix and --output-directory."""
  if not output_directory:
    abs_path = os.path.abspath(input_file)
    release_idx = abs_path.find('Release')
    if release_idx != -1:
      output_directory = abs_path[:release_idx] + 'Release'
      output_directory = os.path.relpath(abs_path[:release_idx] + '/Release')
      logging.debug('Detected --output-directory=%s', output_directory)

  if not tool_prefix and output_directory:
    # Auto-detect from build_vars.txt
    build_vars_path = os.path.join(output_directory, 'build_vars.txt')
    if os.path.exists(build_vars_path):
      with open(build_vars_path) as f:
        build_vars = dict(l.rstrip().split('=', 1) for l in f if '=' in l)
      logging.debug('Found --tool-prefix from build_vars.txt')
      tool_prefix = os.path.join(output_directory,
                                 build_vars['android_tool_prefix'])

  if os.path.sep not in tool_prefix:
    full_path = distutils.spawn.find_executable(tool_prefix + 'c++filt')
  else:
    full_path = tool_prefix + 'c++filt'

  if not full_path or not os.path.isfile(full_path):
    raise Exception('Bad --tool-prefix. Path not found: %s' % full_path)
  if not output_directory or not os.path.isdir(output_directory):
    raise Exception('Bad --output-directory. Path not found: %s' %
                    output_directory)
  logging.info('Using --output-directory=%s', output_directory)
  logging.info('Using --tool-prefix=%s', tool_prefix)
  return output_directory, tool_prefix


def AnalyzeWithArgs(args, input_path):
  return Analyze(input_path, args.output_directory, args.tool_prefix)


def Analyze(path, output_directory=None, tool_prefix=''):
  if path.endswith('.size'):
    logging.debug('Loading results from: %s', path)
    size_info = file_format.LoadSizeInfo(path)
    # Recompute derived values (padding and function names).
    logging.info('Calculating padding')
    _RemoveDuplicatesAndCalculatePadding(size_info.symbols)
    logging.info('Deriving signatures')
    # Re-parse out function parameters.
    _NormalizeNames(size_info.symbols)
    return size_info
  elif not path.endswith('.map') and not path.endswith('.map.gz'):
    raise Exception('Expected input to be a .map or a .size')
  else:
    # Verify tool_prefix early.
    output_directory, tool_prefix = (
        _DetectToolPrefix(tool_prefix, path, output_directory))

    with _OpenMaybeGz(path) as map_file:
      section_sizes, symbols = linker_map_parser.MapFileParser().Parse(map_file)
    timestamp = datetime.datetime.utcfromtimestamp(os.path.getmtime(path))
    size_info = models.SizeInfo(section_sizes, models.SymbolGroup(symbols),
                                timestamp=timestamp)

    # Map file for some reason doesn't unmangle all names.
    logging.info('Calculating padding')
    _RemoveDuplicatesAndCalculatePadding(size_info.symbols)
    # Unmangle prints its own log statement.
    _UnmangleRemainingSymbols(size_info.symbols, tool_prefix)
    logging.info('Extracting source paths from .ninja files')
    _ExtractSourcePaths(size_info.symbols, output_directory)
    # Resolve paths prints its own log statement.
    logging.info('Normalizing names')
    _NormalizeNames(size_info.symbols)
    logging.info('Normalizing paths')
    _NormalizeObjectPaths(size_info.symbols)

  if logging.getLogger().isEnabledFor(logging.INFO):
    for line in describe.DescribeSizeInfoCoverage(size_info):
      logging.info(line)
  logging.info('Finished analyzing %d symbols', len(size_info.symbols))
  return size_info


def _DetectGitRevision(path):
  try:
    git_rev = subprocess.check_output(
        ['git', '-C', os.path.dirname(path), 'rev-parse', 'HEAD'])
    return git_rev.rstrip()
  except Exception:
    logging.warning('Failed to detect git revision for file metadata.')
    return None


def main(argv):
  parser = argparse.ArgumentParser(argv)
  parser.add_argument('input_file', help='Path to input .map file.')
  parser.add_argument('output_file', help='Path to output .size(.gz) file.')
  AddOptions(parser)
  args = helpers.AddCommonOptionsAndParseArgs(parser, argv)
  if not args.output_file.endswith('.size'):
    parser.error('output_file must end with .size')

  size_info = AnalyzeWithArgs(args, args.input_file)
  if not args.input_file.endswith('.size'):
    git_rev = _DetectGitRevision(args.input_file)
    size_info.tag = 'Filename=%s git_rev=%s' % (
        os.path.basename(args.input_file), git_rev)
  logging.info('Recording metadata: %s',
               describe.DescribeSizeInfoMetadata(size_info))
  logging.info('Saving result to %s', args.output_file)
  file_format.SaveSizeInfo(size_info, args.output_file)

  logging.info('Done')


if __name__ == '__main__':
  sys.exit(main(sys.argv))
