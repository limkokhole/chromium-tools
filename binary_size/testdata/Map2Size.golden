Section r has 44.6% of 2641188 bytes accounted for from 7 symbols. 3286464 bytes are unaccounted for. Padding accounts for 14 bytes
+ Without 2 merge sections and 0 anonymous entries (accounting for 2641042 bytes):
+ Section r has 0.0% of 146 bytes accounted for from 5 symbols. 5927506 bytes are unaccounted for. Padding accounts for 14 bytes
Section b has 40.3% of 524520 bytes accounted for from 6 symbols. 775936 bytes are unaccounted for. Padding accounts for 196 bytes
Section d has 0.0% of 388 bytes accounted for from 13 symbols. 1956628 bytes are unaccounted for. Padding accounts for 0 bytes
Section t has 0.0% of 10578 bytes accounted for from 13 symbols. 35890134 bytes are unaccounted for. Padding accounts for 12 bytes
+ Without 2 merge sections and 0 anonymous entries (accounting for 9758 bytes):
+ Section t has 0.0% of 820 bytes accounted for from 11 symbols. 35899892 bytes are unaccounted for. Padding accounts for 16 bytes
.bss@0(size=262144,padding=0,name=ff_cos_131072,path=third_party/fft_float.cc,anon=0)
.bss@0(size=131072,padding=0,name=ff_cos_131072_fixed,path=third_party/fft_fixed.cc,anon=0)
.bss@0(size=131072,padding=0,name=ff_cos_65536,path=third_party/fft_float.cc,anon=0)
.bss@2dffda0(size=28,padding=0,name=g_chrome_content_browser_client,path=third_party/icu/ucnv_ext.c,anon=0)
.bss@2dffe80(size=4,padding=196,name=SaveHistogram::atomic_histogram_pointer,path=third_party/icu/ucnv_ext.c,anon=0)
.bss@2dffe84(size=4,padding=0,name=g_AnimationFrameTimeHistogram_clazz,path=third_party/icu/ucnv_ext.c,anon=1)
.data@2de7000(size=4,padding=0,name=google::protobuf::internal::pLinuxKernelCmpxchg,path=base/page_allocator.cc,anon=0)
.data@2de7004(size=4,padding=0,name=google::protobuf::internal::pLinuxKernelMemoryBarrier,path=third_party/container.c,anon=0)
.data@2de7008(size=152,padding=0,name=rel._ZN4base7androidL22kBaseRegisteredMethodsE,path=third_party/container.c,anon=0)
.data@2de70a0(size=4,padding=0,name=base::android::g_renderer_histogram_code,path=third_party/container.c,anon=1)
.data@2de70a4(size=4,padding=0,name=rel.local._ZN4base7android12_GLOBAL__N_124g_library_version_numberE,path=third_party/container.c,anon=0)
.data.rel.ro@2cd8500(size=56,padding=0,name=ChromeMainDelegateAndroid [vtable],path=third_party/paint.cc,anon=0)
.data.rel.ro@2cd8538(size=24,padding=0,name=mojo::MessageReceiver [vtable],path=base/page_allocator.cc,anon=0)
.data.rel.ro@2cd8550(size=12,padding=0,name=kMethodsAnimationFrameTimeHistogram,path=base/page_allocator.cc,anon=0)
.data.rel.ro.local@2c176f0(size=56,padding=0,name=ChromeMainDelegate [vtable],path=third_party/icu/ucnv_ext.c,anon=0)
.data.rel.ro.local@2c17728(size=24,padding=0,name=chrome::mojom::FieldTrialRecorderRequestValidator [vtable],path=third_party/icu/ucnv_ext.c,anon=0)
.data.rel.ro.local@2c17740(size=24,padding=0,name=chrome::mojom::FieldTrialRecorderProxy [vtable],path=third_party/container.c,anon=0)
.data.rel.ro.local@2cd84e0(size=16,padding=0,name=.Lswitch.table.45,path=third_party/gvr-android-sdk/libgvr_shim_static_arm.a/(libcontroller_api_impl.a_controller_api_impl.o),anon=0)
.data.rel.ro.local@2cd84f0(size=8,padding=0,name=kSystemClassPrefixes,path=third_party/gvr-android-sdk/libgvr_shim_static_arm.a/(libport_android_jni.a_jni_utils.o),anon=1)
.rodata@266e600(size=1965409,padding=0,name=** merge strings,path=,anon=0)
.rodata@284e364(size=8,padding=3,name=,path=base/page_allocator.cc,anon=0)
.rodata@284e370(size=40,padding=4,name=Name,path=base/page_allocator.cc,anon=0)
.rodata@284e398(size=32,padding=0,name=chrome::mojom::FilePatcher::Name_,path=third_party/container.c,anon=0)
.rodata@284e518(size=675633,padding=0,name=** merge strings,path=,anon=0)
.rodata@28f3450(size=48,padding=7,name=kAnimationFrameTimeHistogramClassPath,path=third_party/paint.cc,anon=1)
.rodata@28f3480(size=4,padding=0,name=blink::CSSValueKeywordsHash::findValueImpl::value_word_list,path=third_party/paint.cc,anon=1)
.text@28d900(size=16,padding=0,name=startup._GLOBAL__sub_I_page_allocator.cc,path=base/page_allocator.cc,anon=0)
.text@28d910(size=56,padding=0,name=startup._GLOBAL__sub_I_bbr_sender.cc,path=base/page_allocator.cc,anon=0)
.text@28d948(size=28,padding=0,name=startup._GLOBAL__sub_I_pacing_sender.cc,path=base/page_allocator.cc,anon=0)
.text@28d964(size=38,padding=0,name=extFromUUseMapping,path=base/page_allocator.cc,anon=0)
.text@28d98a(size=32,padding=0,name=extFromUUseMapping,path=base/page_allocator.cc,anon=0)
.text@28d9aa(size=5718,padding=0,name=** symbol gap 0,path=third_party/icu/ucnv_ext.c,anon=0)
.text@28f000(size=448,padding=0,name=ucnv_extMatchFromU,path=third_party/icu/ucnv_ext.c,anon=0)
.text@28f1c8(size=20,padding=8,name=startup._GLOBAL__sub_I_SkDeviceProfile.cpp,path=third_party/icu/ucnv_ext.c,anon=0)
.text@28f1e0(size=20,padding=4,name=startup._GLOBAL__sub_I_SkDiscardableMemoryPool.cpp,path=third_party/icu/ucnv_ext.c,anon=0)
.text@2a0000(size=28,padding=0,name=blink::ContiguousContainerBase::shrinkToFit,path=third_party/paint.cc,anon=0)
.text@2a0020(size=24,padding=4,name=blink::ContiguousContainerBase::ContiguousContainerBase,path=third_party/container.c,anon=0)
.text@2a0034(size=4044,padding=-4,name=** symbol gap 1,path=third_party/container.c,anon=0)
.text@2a1000(size=94,padding=0,name=blink::PaintChunker::releasePaintChunks,path=third_party/container.c,anon=1)
